std/container/rbtree.d

```diff
@@ -1094,6 +1094,108 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
         assert(t1 != o);  // pathological case, must not crash
     }
 
+    /**
+     * Generates a hash for the tree
+     */
+    override size_t toHash() nothrow @safe
+    {
+        import std.traits : hasMember;
+
+        auto thisRange = this[];
+
+        static if (hasMember!(Elem, "toHash"))
+        {
+            size_t hash = 0;
+
+            foreach (ref e; thisRange)
+            {
+                hash += e.toHash() + 0x9e3779b9 + (hash << 6) + (hash >>> 2);
```

Comment:
If `is(T == class)` instead of `e.toHash()` we need to write `(e is null ? 0 : e.toHash())`, unless we know the tree disallows null elements.

Reply:
How can I check if the tree disallows null elements? I have made the suggested changes, and I have written this test:
```
        class MyInt2
        {
            int x;

            this(int init_x)
            {
                x = init_x;
            }

            override size_t toHash() const @safe nothrow
            {
                return typeid(x).getHash(&x) ^ 0xF0F0_F0F0;
            }

            int opCmp(const MyInt2 that) const
            {
                return (x > that.x) - (x < that.x);
            }

            bool opEquals(const MyInt2 that) const
            {
                return (this.x == that.x);
            }
        }

        auto rbt4 = new RedBlackTree!MyInt2(null); // this is okay
        auto rbt5 = new RedBlackTree!MyInt2(new MyInt2(1), null, new MyInt2(3), null); // this generates a sigsev
```

The output is:
```
/tmp/dmd_run5P9L1u(+0x1eaa8c)[0x5642324f3a8c]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x110c0)[0x7f5ef63b10c0]
/tmp/dmd_run5P9L1u(+0x6a5d2)[0x5642323735d2]
/tmp/dmd_run5P9L1u(+0x67e37)[0x564232370e37]
/tmp/dmd_run5P9L1u(+0x6a56c)[0x56423237356c]
/tmp/dmd_run5P9L1u(+0x691ce)[0x5642323721ce]
/tmp/dmd_run5P9L1u(+0x51d60)[0x56423235ad60]
/tmp/dmd_run5P9L1u(+0x1da1ae)[0x5642324e31ae]
/tmp/dmd_run5P9L1u(+0x1eaae0)[0x5642324f3ae0]
/tmp/dmd_run5P9L1u(+0x1dc53f)[0x5642324e553f]
/tmp/dmd_run5P9L1u(+0x1e2c7e)[0x5642324ebc7e]
/tmp/dmd_run5P9L1u(+0x1e2e11)[0x5642324ebe11]
/tmp/dmd_run5P9L1u(+0x1e2c0a)[0x5642324ebc0a]
/tmp/dmd_run5P9L1u(+0x1dc516)[0x5642324e5516]
/tmp/dmd_run5P9L1u(+0x1ea8b6)[0x5642324f38b6]
/tmp/dmd_run5P9L1u(+0x1ddd85)[0x5642324e6d85]
/tmp/dmd_run5P9L1u(+0x1ddd0c)[0x5642324e6d0c]
/tmp/dmd_run5P9L1u(+0x1ddc77)[0x5642324e6c77]
/tmp/dmd_run5P9L1u(+0x1da9b4)[0x5642324e39b4]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7f5ef58fe2e1]
/tmp/dmd_run5P9L1u(+0x4d5ea)[0x5642323565ea]
Error: program killed by signal 11
posix.mak:399: recipe for target 'std/container/rbtree.test' failed
make: *** [std/container/rbtree.test] Error 1
```

Reply:
>How can I check if the tree disallows null elements?

In principle we could tell this by seeing if the comparison predicate accepted null values, but I think it's fine to just assume that if `is(T == class) || is(T == interface)` then `null` is a possibility.

>I have made the suggested changes
>Error: program killed by signal 11

It's segfaulting because you still need to do:
```d
hash += (e is null ? 0 : e.toHash()) + 0x9e3779b9 + (hash << 6) + (hash >>> 2);
```
instead of 
```d
hash += e.toHash() + 0x9e3779b9 + (hash << 6) + (hash >>> 2);
```
You don't need to do the same in the other branch.